### PR TITLE
Fixes #12692: Update samples/gpu/videoreader.cpp

### LINKED ISSUE #12692: Error in cuda videoreader sample program
##### System information (version)
- OpenCV => 3.4
- Cuda - 9.2
- Operating System / Platform => Ubuntu 16.04 64 Bit
- Compiler => g++ 4.9

##### Detailed description
The videoreader example has both the cpu and gpu reading of video frames in the same for loop. This should be moved to for loops as the GPU decoded video could have more number of frames.

I am not sure if this is a bug in cuda videoreader code or in the sample program.
##### Steps to reproduce

Moving the vVideoCapture  to different loop and adding a frame counter gives different results

`Sample Output`

```
Results:
CPU : Avg : 4.82522 ms FPS : 207.245 1794

GPU : Avg : 0.549416 ms FPS : 1820.11 3572

```

### PR DESCRIPTION
**Pull request by sam09**
Change iteration to iterate separately when the video is decode on a GPU to account for different number of frames

resolves #12692 

```
force_builders=linux,docs,Custom
docker_image:Custom=ubuntu-cuda:16.04
```

### CODE REVIEW

File: samples/gpu/video_reader.cpp [hunk 1]
```diff
@@ -35,21 +35,32 @@ int main(int argc, const char* argv[])
...
+    }
 
+    for (;;)
+    {
         tm.reset(); tm.start();
         if (!d_reader->nextFrame(d_frame))
             break;
         tm.stop();
         gpu_times.push_back(tm.getTimeMilli());
+        cpu_frame_count++;
```
Comment (alalek) [inline samples/gpu/video_reader.cpp]: gpu_frame_count

File: samples/gpu/video_reader.cpp [hunk 2]
```diff
@@ -66,8 +77,8 @@ int main(int argc, const char* argv[])
         double cpu_avg = std::accumulate(cpu_times.begin(), cpu_times.end(), 0.0) / cpu_times.size();
         double gpu_avg = std::accumulate(gpu_times.begin(), gpu_times.end(), 0.0) / gpu_times.size();
 
-        std::cout << "CPU : Avg : " << cpu_avg << " ms FPS : " << 1000.0 / cpu_avg << std::endl;
-        std::cout << "GPU : Avg : " << gpu_avg << " ms FPS : " << 1000.0 / gpu_avg << std::endl;
+        std::cout << "CPU : Avg : " << cpu_avg << " ms FPS : " << 1000.0 / cpu_avg << " Frames" << cpu_frame_count <<std::endl;
+        std::cout << "GPU : Avg : " << gpu_avg << " ms FPS : " << 1000.0 / gpu_avg << " Frames" << gpu_frame_count << std::endl;
```
[the remark this answers is not in the document]
Comment (sam09) [reply samples/gpu/video_reader.cpp]: Yes I did. I ran with a test video which gave me different results for cpu_frame_count and gpu_frame_count. Not sure why this happens.

[the remark this answers is not in the document]
Comment (alalek) [reply samples/gpu/video_reader.cpp]: This is strange, but we observed total fps count +/- 1 frame difference between different codecs / versions (and we can't fix that in OpenCV).

I mean missing space after "Frames":

> " Frames" << cpu_frame_count

=> " Frames : "

[the remark this answers is not in the document]
Comment (sam09) [reply samples/gpu/video_reader.cpp]: Ah. Okay.
I also saw almost twice the number of frames when decoding using GPU which is very strange.

**Pull request merged**